R/db_connections.R

```diff
@@ -1,263 +1,157 @@
-    colNumsDTS <- grep("DTS", names(df))
-    df[colNumsDTS] <- lapply(df[colNumsDTS], as.POSIXct)
+  if (is.character(database)) {
+    con_str <- paste0(con_str,
+                      "database=", database, ";")
   }
 
-  # Make sure there are enough rows to actually do something useful.
-  if (is.null(nrow(df))) {
-    cat(df)  # Print the SQL error, which is contained in df.
-    stop("Your SQL contains an error.")
-  }
-  if (nrow(df) == 0) {
-    warning("Zero rows returned from SQL. ",
-            "Adjust your query to return more data!")
+  if (rlang::is_true(trusted)) {
+    con_str <- paste0(con_str,
+                      "trusted_connection=true;")
+  } else if (!is.na(user_id) & !is.na(password)) {
+    con_str <- paste0(con_str,
+                      "uid=", user_id, ";",
+                      "pwd=", password, ";")
+  } else {
+    stop("You must use trusted=TRUE or provide a user_id and password.")
   }
-  df  # Return the selected data.
+  return(con_str)
 }
 
 #' @title
-#' Write data to database
-#' @description Write data frame to database table via ODBC connection
-#' @param MSSQLConnectionString A string specifying the driver, server,
-#' database, and whether Windows Authentication will be used.
-#' @param df Dataframe that hold the tabular data
-#' @param SQLiteFileName A string. If dbtype is SQLite, here one specifies the
-#' database file to query from
-#' @param tableName String. Name of the table that receives the new rows
-#' @param addSAMUtilityColumns Boolean. Whether to add Health Catalyst-related
-#' date-time stamp, BindingID, and BindingNM to df before saving to db.
-#' @param connectionString Deprecated. A string specifying the driver, server,
-#' database, and whether Windows Authentication will be used. See
-#' ?MSSQLConnectionString instead.
-#' @return Nothing
+#' Read from a SQL Server database table
+#' @description Use a database connection to read from an existing SQL Server
+#' table with a SQL query.
+#' @param con An odbc database connection. Can be made using
+#' \code{build_connection_string}. Required.
+#' @param query A string, quoted, required. This sql query will be executed
+#' against the database you are connected to.
+#' @param pull_into_memory Logical, optional, defaults to TRUE. If FALSE,
+#' \code{db_read} will create a reference to the queried data rather than
+#' pulling into memory.
+#' @details Use \code{pull_into_memory} when working with large tables.
+#' Rather than returning the data into memory, this
+#' function will return a reference to the specified query. It will be executed
+#' only when needed, in a "lazy" style. Or, you can execute using the
+#' \code{collect()} function.
+#' @return A tibble of data or reference to the table.
+#' @seealso \code{\link{build_connection_string}}
 #' @export
-#' @references \url{http://healthcareai-r.readthedocs.io}
-#' @seealso \code{\link{healthcareai}}
-#' @seealso \code{\link{selectData}}
 #' @examples
```

Comment:
Let's use a realistic example. Here's what I used to test on HPH if you want to just use this go ahead:

my_con <- build_connection_string(server = "HPHI-EDWDEV")
con <- DBI::dbConnect(odbc::odbc(), .connection_string = my_con)
d <- db_read(con, "SELECT TOP 10 * FROM [Shared].[Cost].[FacilityAccountCost]")

Reply:
Done, switched this out. How self serving! I think it's better this way.